File: packages/circuitpy/src/circuitpy/netwidth.py
"""How wide a net can be, and how wide it is — the two numbers a human needs
before they touch a trace width.

EE review 2026-08-15, finding 4: *"the main power nets, 5V and 3V3, are
currently the same width as a signal, 0.2mm ... should be 0.5–1.0mm"*. The
board file can say so — ``<trace thickness="0.5mm">`` reaches the autorouter as
a per-net ``nominalTraceWidth`` it holds while it searches — and saying it
blindly scraps the board. Measured twice, on two different days:

* harness-puck, every rail declared at once: ``fab.ready`` **true → false**,
  0 blocking findings → **33**, every one of them inside the RP2040's fanout.
* terminal-keyboard, V3_3 alone declared at 0.5mm from the board file:
  0 blocking → **2**, one of them ``Track [V3_3] ... [clearance] 0.0501mm``
  against the 0.0900mm floor.

Neither is a router failure. It is arithmetic: a track leaving a QFN-56 pad at
0.400mm pitch with 0.200mm pads can be
``2 × (0.400 − 0.100 − 0.100) = 0.400mm`` wide and no wider, so a 0.5mm rail
through that fanout cannot exist at any effort, with any router.

So the tool's job is not to widen rails. It is to **hand the engineer the
ceiling before they choose**, per net:

``ceiling_mm``
    the widest track that can leave the net's own pads, measured against the
    *placement* — pads, plated holes, NPTH, keepouts, board edge. Traces, vias
    and pours are excluded on purpose: they are the router's output, and the
    question is what the placement permits, not what this route happened to do.

``narrowest_mm``
    what the net is actually routed at today, at its worst point, which is what
    ``dfm_power_trace_width`` reports and therefore what a fab sees.

``declared_mm``
    what the board file asked for, if anything.

The measurement is the one from ``work/railwidth/escape.py``, which is where
the numbers in ``docs/architecture/rail-width.md`` came from; this is that
script with a return value instead of a print, so the app can show it.
"""

from __future__ import annotations

import math
from typing import Any, Sequence

from circuitpy import diffpair, powerwidth

#: How far out of the pad the escape has to survive, in mm. Past the far edge
#: of the neighbouring pads the channel opens up, so a shorter probe flatters
#: the escape; 1.0mm clears a 0.85mm QFN pad end to end.
PROBE_MM = 1.0

#: Bearings tried per pad. 180 is one every two degrees, which is finer than
#: any pad pitch we ship.
DIRECTIONS = 180

#: The widest answer this asks for, in mm.
#:
#: It is a speed knob and an honesty knob at once. `_segment_limit`'s
#: bounding-box prefilter is sized from the cap (`cap / 2 + 1.0`), so the 50mm
#: the one-off script passed meant a 26mm reach and effectively no prefilter at
#: all: 11.3 seconds for two rails on terminal-keyboard, which is not a number
#: you can put behind a click. At 2.0mm the prefilter does its job and the same
#: two rails take a fraction of that.
#:
#: 2.0mm is past anything a board of ours would route — JLC's own sample rules
#: and every rail in `circuitlib` live under 1.0mm — and a pad that can escape
#: wider than the cap is reported **at** the cap with `ceiling_capped`, which
#: reads "at least this", never "exactly this".
CAP_MM = 2.0

#: Ceilings are reported to the micrometre. The escape is a measurement, and
#: printing it to fourteen decimals in a UI is how 0.4000000000000001 ends up
#: in front of an engineer.
_ROUND = 4


def escape_width(
    cx: float,
    cy: float,
    layer: str,
    net_key: str | None,
    obstacles: Sequence[Any],
    outline: Any,
    edge: float,
    cap: float = CAP_MM,
) -> tuple[float, float, str | None]:
    """The widest track that can leave ``(cx, cy)``, its bearing, and what held
    it there."""
    best = -math.inf
    bearing = 0.0
    limiter: str | None = None
    for i in range(DIRECTIONS):
        theta = math.pi * i / DIRECTIONS * 2
        seg = (cx, cy, cx + PROBE_MM * math.cos(theta), cy + PROBE_MM * math.sin(theta))
        allowed, who = powerwidth._segment_limit(
            seg, layer, net_key, obstacles, outline, edge, cap
        )
        if allowed > best:
            best, bearing, limiter = allowed, math.degrees(theta), who
    return best, bearing, limiter


def _placement_obstacles(board: Any, elements: Sequence[dict], profile: Any) -> list:
    """Everything the *placement* puts in a track's way.

    Not traces (a `capsule`), not vias, not pours — those are the router's
    output, and a ceiling measured against them would move every time the
    router did. Pour ids are net-derived strings rather than anything
    containing the word "pour", so they are collected from the elements; routed
    copper is filtered by *shape* for the same reason.
    """
    pour_labels = {
        str(e.get("pcb_copper_pour_id") or "")
        for e in elements
        if isinstance(e, dict) and e.get("type") == "pcb_copper_pour"
    }
    return [
        o
        for o in diffpair.collect_obstacles(board, profile, set())
        if o.kind != "capsule"
        and not o.label.startswith("pcb_via")
        and o.label not in pour_labels
    ]


def net_width_report(
    elements: Sequence[dict],
    profile: Any,
    *,
    nets: Sequence[str] | None = None,
    rails_only: bool = False,
) -> list[dict]:
    """One row per net, newest measurement first in the caller's order.

    ``nets`` selects by name (the net's own name, as the board file spells it);
    ``rails_only`` selects the power and ground nets `powerwidth` recognises.
    With neither, every net that has at least one SMD pad is measured — which
    on terminal-keyboard is 300+ pads and takes seconds, so a caller with a
    person waiting should ask for the net they are looking at.
    """
    board = diffpair._Board(elements)
    outline = diffpair._board_outline(board)
    edge = float(profile.min_edge_clearance_mm)
    placement = _placement_obstacles(board, elements, profile)

    rails = powerwidth._rail_source_nets(elements)
    wanted = {str(n) for n in nets} if nets else None

    components = {
        e["source_component_id"]: e.get("name")
        for e in elements
        if isinstance(e, dict) and e.get("type") == "source_component"
    }
    source_ports = {
        e["source_port_id"]: e
        for e in elements
        if isinstance(e, dict) and e.get("type") == "source_port"
    }
    routed = _narrowest_by_net(elements, board)
    declared = declared_widths(elements)

    rows: list[dict] = []
    for net in board.source_nets.values():
        name = str(net.get("name") or "")
        nid = str(net.get("source_net_id") or "")
        key = net.get("subcircuit_connectivity_map_key")
        if not name or not isinstance(key, str):
            continue
        if wanted is not None and name not in wanted:
            continue
        if rails_only and nid not in rails:
            continue

        pads: list[dict] = []
        for pad in board.by_type.get("pcb_smtpad", []):
            port_id = str(pad.get("pcb_port_id") or "")
            if board.net_key_of_pcb_port(port_id) != key:
                continue
            if pad.get("x") is None or pad.get("y") is None:
                # A polygon pad carries `points` and no centre — the USB-C
                # shell pads are the only ones we ship. Counted, not guessed.
                pads.append({"pad": str(pad.get("pcb_smtpad_id") or ""), "skipped": True})
                continue
            port = board.pcb_ports.get(port_id) or {}
            sp = source_ports.get(str(port.get("source_port_id") or "")) or {}
            width, bearing, limiter = escape_width(
                float(pad["x"]),
                float(pad["y"]),
                str(pad.get("layer") or "top"),
                key,
                placement,
                outline,
                edge,
            )
            pads.append(
                {
                    "pad": str(pad.get("pcb_smtpad_id") or ""),
                    "port": f"{components.get(sp.get('source_component_id'), '?')}."
                    f"{sp.get('name', '?')}",
                    "escape_mm": round(max(width, 0.0), _ROUND),
                    "capped": width >= CAP_MM - 1e-9,
                    "bearing_deg": round(bearing, 1),
                    "limiter": limiter,
                }
            )

        measured = [p for p in pads if "escape_mm" in p]
        if not measured:
            continue
        tightest = min(measured, key=lambda p: p["escape_mm"])
        rows.append(
            {
                "net": name,
                "rail": nid in rails,
                # The ceiling is the *tightest* pad, because a rail is only as
                # wide as its worst escape — the same rule
                # `dfm_power_trace_width` reports the board by.
                "ceiling_mm": tightest["escape_mm"],
                "ceiling_at": tightest["port"],
                # "at least this wide" — the measurement stopped at the cap
                # rather than finding a limit.
                "ceiling_capped": bool(tightest.get("capped")),
                "ceiling_limiter": tightest["limiter"],
                "pads": len(measured),
                "pads_skipped": len(pads) - len(measured),
                "narrowest_mm": routed.get(key),
                "declared_mm": declared.get(name),
            }
        )
    return rows


def _narrowest_by_net(elements: Sequence[dict], board: Any) -> dict[str, float]:
    """The thinnest copper each net is actually routed at, keyed by net key."""
    out: dict[str, float] = {}
    for element in elements:
        if not isinstance(element, dict) or element.get("type") != "pcb_trace":
            continue
        key = None
        for point in element.get("route") or []:
            port_id = point.get("start_pcb_port_id") or point.get("end_pcb_port_id")
            if port_id:
                key = board.net_key_of_pcb_port(str(port_id))
                if key:
                    break
        if not key:
            continue
        for point in element.get("route") or []:
            width = point.get("width")
            if width is None:
                continue
            try:
                value = float(width)
            except (TypeError, ValueError):
                continue
            out[key] = min(out.get(key, value), value)
    return {k: round(v, _ROUND) for k, v in out.items()}


def declared_widths(elements: Sequence[dict]) -> dict[str, float]:
    """What the board file asked for, per net name.

    `getSimpleRouteJsonFromCircuitJson` gives a net-shaped connection
    ``nominalTraceWidth = max(min_trace_thickness)`` over every trace joined to
    that net, so **one declaration anywhere on a net sets the whole net** and a
    partial marking is identical to a full one. The max is why this reports the
    largest declaration rather than the first.
    """
    nets_by_id = {
        str(e.get("source_net_id") or ""): str(e.get("name") or "")
        for e in elements
        if isinstance(e, dict) and e.get("type") == "source_net"
    }
    out: dict[str, float] = {}
    for element in elements:
        if not isinstance(element, dict) or element.get("type") != "source_trace":
            continue
        thickness = element.get("min_trace_thickness")
        if thickness is None:
            continue
        try:
            value = float(thickness)
        except (TypeError, ValueError):
            continue
        for net_id in element.get("connected_source_net_ids") or []:
            name = nets_by_id.get(str(net_id))
            if name:
                out[name] = max(out.get(name, value), value)
    return {k: round(v, _ROUND) for k, v in out.items()}

File: viewer/src/client/components/board/netWidth.js
// netWidth — declaring how wide a net is routed, as one edit to the board file.
//
// The EE review's finding 4, in the engineer's own words: *"the main power
// nets, 5V and 3V3, are currently the same width as a signal, 0.2mm ... should
// be 0.5–1.0mm"*. The board file can say so and the router obeys it —
// `<trace thickness="0.5mm">` reaches the autorouter as a per-net
// `nominalTraceWidth` it holds while it searches — so this is a real edit a
// human can make, not a request an agent has to translate.
//
// No DOM, no React, no `@/` alias, for the reason `boardSource.js` has none:
// the part that can be wrong invisibly is the text arithmetic, and `node:test`
// covers it directly.
//
// Two facts from the shipped bundle decide the whole design:
//
//   1. **One declaration anywhere on a net sets the whole net.**
//      `getSimpleRouteJsonFromCircuitJson` takes
//      `nominalTraceWidth = max(min_trace_thickness)` over every trace joined
//      to the net. So marking one trace is identical to marking all of them,
//      and this writes exactly one.
//   2. **A board-level trace is enough**, even when the net is wired inside a
//      block. Measured 2026-08-16 on terminal-keyboard: adding
//      `<trace from=".U2 > .VOUT" to="net.V3_3" thickness="0.5mm" />` to the
//      board file moved V3_3's copper (and cost the board two clearance
//      findings inside the RP2040 fanout, which is the ceiling doing its job).
//
// What this module will not do is decide the number. The ceiling comes from
// `circuitpy.netwidth`, measured against the placement, and the caller shows
// it: 0.4000mm on V3_3 through a QFN-56's 0.400mm pitch, 1.1mm on V5. A tool
// that let someone type 0.5 into the first of those without saying so would be
// handing them a board that cannot exist.

import { maskCommentsAndStrings } from "./boardSource.js";

/** How a width is spelled in the file. Millimetres, the board's own unit. */
export function formatWidth(mm) {
  const value = Number(mm);
  if (!Number.isFinite(value) || value <= 0) return "";
  // Trailing zeros stripped: `0.5mm`, not `0.500mm`. Four decimals is the
  // micrometre, which is finer than any fab tolerance we quote.
  return `${String(Number(value.toFixed(4)))}mm`;
}

/**
 * Every `<trace …/>` in the file, with its span and its text.
 *
 * `maskCommentsAndStrings` returns a **mask**, one byte per character, 1 where
 * that character is inside a comment or a string — not a masked copy of the
 * text. Both the opening `<trace` and the `>` that ends the tag are only real
 * where the mask is 0, which is what keeps `from=".U3 > .GPIO0"` from ending
 * the tag at its own greater-than and a `<trace …/>` written inside a comment
 * from being edited at all.
 */
function traceTags(source) {
  const text = String(source || "");
  const mask = maskCommentsAndStrings(text);
  const tags = [];
  const re = /<trace\b/g;
  let match;
  while ((match = re.exec(text)) !== null) {
    const start = match.index;
    if (mask[start]) continue;
    let end = -1;
    for (let i = start; i < text.length; i += 1) {
      if (text[i] === ">" && !mask[i]) {
        end = i;
        break;
      }
    }
    if (end === -1) break;
    tags.push({ start, end: end + 1, text: text.slice(start, end + 1) });
    re.lastIndex = end + 1;
  }
  return tags;
}

/** The net a `to=`/`from=` value names, or "" — `net.V3_3` → `V3_3`. */
function netOf(value) {
  const match = /^net\.([A-Za-z0-9_]+)$/.exec(String(value || "").trim());
  return match ? match[1] : "";
}

/** `attr="value"` out of one tag's text. */
function attr(text, name) {
  const match = new RegExp(`\\b${name}=(?:"([^"]*)"|\\{\`([^\`]*)\`\\})`).exec(text);
  return match ? match[1] ?? match[2] ?? "" : "";
}

/**
 * The traces in this file that land on `net`, nearest thing to a handle the
 * board has on that net's width.
 */
export function tracesForNet(source, net) {
  const wanted = String(net || "");
  if (!wanted) return [];
  return traceTags(String(source || "")).filter(
    (tag) => netOf(attr(tag.text, "to")) === wanted || netOf(attr(tag.text, "from")) === wanted,
  );
}

/** What the file currently declares for `net`, in mm, or null. */
export function declaredWidth(source, net) {
  let best = null;
  for (const tag of tracesForNet(source, net)) {
    const raw = attr(tag.text, "thickness");
    const value = Number(String(raw).replace(/mm$/i, ""));
    if (Number.isFinite(value) && value > 0) best = best === null ? value : Math.max(best, value);
  }
  return best;
}

/**
 * The edits that make `net` route at `mm`, or a refusal saying why not.
 *
 * `mm` of 0 or null clears the declaration, which is how a rail goes back to
 * the board's own `minTraceWidth` — an engineer who widens a net and does not
 * like the result must be able to put it back without hand-editing TSX.
 *
 * @returns {{ok: true, edits: Array, summary: string} | {ok: false, reason: string}}
 */
export function widthEdits(source, net, mm, { anchorPort = "" } = {}) {
  const text = String(source || "");
  const name = String(net || "").trim();
  if (!name) return { ok: false, reason: "no net named" };
  const want = mm === null || mm === undefined || Number(mm) === 0 ? null : Number(mm);
  if (want !== null && (!Number.isFinite(want) || want <= 0 || want > 10)) {
    return { ok: false, reason: `${want}mm is not a track width` };
  }

  const tags = tracesForNet(text, name);
  const current = declaredWidth(text, name);
  if (current === want || (current === null && want === null)) {
    return { ok: false, reason: `${name} is already ${want === null ? "undeclared" : formatWidth(want)}` };
  }

  // The trace that already carries the declaration, else the first one on the
  // net — one is enough, and marking a second would leave two numbers to
  // disagree later.
  const marked = tags.find((tag) => attr(tag.text, "thickness")) || tags[0] || null;

  if (marked) {
    const existing = /\s*thickness=(?:"[^"]*"|\{`[^`]*`\})/.exec(marked.text);
    if (want === null) {
      if (!existing) return { ok: false, reason: `${name} has no declared width to clear` };
      const start = marked.start + existing.index;
      return {
        ok: true,
        summary: `${name} back to the board's default width`,
        edits: [{ start, end: start + existing[0].length, text: "", expected: existing[0] }],
      };
    }
    if (existing) {
      const start = marked.start + existing.index;
      return {
        ok: true,
        summary: `${name} routed at ${formatWidth(want)}`,
        edits: [
          {
            start,
            end: start + existing[0].length,
            text: ` thickness="${formatWidth(want)}"`,
            expected: existing[0],
          },
        ],
      };
    }
    // No `thickness` yet: insert one after the last attribute rather than
    // immediately before the close, or a tag written `… />` comes back
    // `…  thickness="0.5mm"/>` — two spaces and none where the eye expects
    // one. The file is read by people.
    const closeLen = marked.text.endsWith("/>") ? 2 : 1;
    const body = marked.text.slice(0, marked.text.length - closeLen);
    const closeAt = marked.start + body.replace(/\s+$/, "").length;
    return {
      ok: true,
      summary: `${name} routed at ${formatWidth(want)}`,
      edits: [{ start: closeAt, end: closeAt, text: ` thickness="${formatWidth(want)}"`, expected: "" }],
    };
  }

  // No trace in the board file touches this net — it is wired inside a block.
  // A new board-level trace between a port on the net and the net is a
  // connection that already exists, so it adds no copper; what it adds is the
  // `min_trace_thickness` the router reads. Measured working on
  // terminal-keyboard's V3_3, which is wired entirely inside `ldo-3v3`.
  if (want === null) return { ok: false, reason: `${name} has no declared width to clear` };
  if (!anchorPort) {
    return {
      ok: false,
      reason:
        `${name} is wired inside a block, so the board file has no trace to widen — ` +
        "and no pin on the net was given to hang one on",
    };
  }
  const closing = text.lastIndexOf("</board>");
  if (closing === -1) return { ok: false, reason: "this board file has no <board> to add a trace to" };
  const indent = "    ";
  const insert =
    `${indent}{/* width: ${name} routed at ${formatWidth(want)} (set in the app) */}\n` +
    `${indent}<trace from="${anchorPort}" to="net.${name}" thickness="${formatWidth(want)}" />\n`;
  return {
    ok: true,
    summary: `${name} routed at ${formatWidth(want)}`,
    edits: [{ start: closing, end: closing, text: insert, expected: "" }],
  };
}

/**
 * A pin on this net that the board file can name, as a tscircuit selector.
 *
 * Only needed when the net is wired entirely inside a block: the board file
 * then has no trace to mark, and a board-level trace between one of the net's
 * own pins and the net gives the router the width without adding copper,
 * because that connection already exists. `.U2 > .VOUT` is the spelling
 * tscircuit uses and the one every trace in our board files already carries.
 *
 * Read off the built board rather than the source, because the source is
 * exactly what does not mention this net.
 */
export function anchorPortForNet(index, netKey) {
  const elements = index?.elements;
  if (!index || !netKey || !Array.isArray(elements)) return "";
  const sourcePorts = new Map();
  const refdesById = new Map();
  for (const element of elements) {
    if (!element || typeof element !== "object") continue;
    if (element.type === "source_port") sourcePorts.set(String(element.source_port_id), element);
    else if (element.type === "source_component") {
      refdesById.set(String(element.source_component_id), String(element.name || ""));
    }
  }
  const net = index.netByKey?.get(netKey);
  for (const id of net?.pcbElementIds || []) {
    const element = index.byId?.get(id);
    if (!element || (element.type !== "pcb_smtpad" && element.type !== "pcb_plated_hole")) continue;
    const pcbPort = index.byId?.get(String(element.pcb_port_id || ""));
    const source = sourcePorts.get(String(pcbPort?.source_port_id || ""));
    if (!source) continue;
    const refdes = refdesById.get(String(source.source_component_id || ""));
    const pin = String(source.name || "");
    // A pin with no name is a pin the file cannot address; keep looking.
    if (refdes && pin) return `.${refdes} > .${pin}`;
  }
  return "";
}
